File: tedorilog-poc/engine/layout.js
// トークン列 → 行・列構造への整形。
// PDF直接抽出とOCRで粒度が違う（OCRは単語が細かく割れる）ため、
// 一度ここで「行」「連結トークン」へ揃えてから解析する。

import { parseAmount, toHalfWidth } from './normalize.js';

const clone = (t) => ({ ...t });

function tokenRight(t) {
  return t.x + (t.w || 0);
}

function tokenCenterY(t) {
  return t.y + (t.h || 0) / 2;
}

function tokenCenterX(t) {
  return t.x + (t.w || 0) / 2;
}

export function medianOf(values) {
  if (!values.length) return 0;
  const s = [...values].sort((a, b) => a - b);
  return s[Math.floor(s.length / 2)];
}

/** 与えられた傾き slope のもとで行キーを計算する（左上原点、slopeは dy/dx）。 */
function rowKey(t, slope) {
  return tokenCenterY(t) - slope * tokenCenterX(t);
}

function clusterRows(tokens, slope, tolerance) {
  const sorted = [...tokens].sort((a, b) => rowKey(a, slope) - rowKey(b, slope));
  const rows = [];
  let current = [];
  let anchor = null;
  for (const t of sorted) {
    const k = rowKey(t, slope);
    if (anchor === null || Math.abs(k - anchor) <= tolerance) {
      if (anchor === null) anchor = k;
      current.push(t);
      anchor = (anchor * (current.length - 1) + k) / current.length;
    } else {
      rows.push(current);
      current = [t];
      anchor = k;
    }
  }
  if (current.length) rows.push(current);
  return rows;
}

/**
 * 紙写真の傾きを推定する。行クラスタ数と行内のばらつきが最小になる傾きを選ぶ。
 * @returns {number} slope (dy/dx)
 */
export function estimateSkew(tokens, medianHeight) {
  if (tokens.length < 6) return 0;
  const tolerance = Math.max(2, medianHeight * 0.5);
  let best = { slope: 0, cost: Infinity };
  for (let slope = -0.08; slope <= 0.0801; slope += 0.0025) {
    const rows = clusterRows(tokens, slope, tolerance);
    let spread = 0;
    for (const row of rows) {
      const keys = row.map((t) => rowKey(t, slope));
      spread += Math.max(...keys) - Math.min(...keys);
    }
    const cost = rows.length * tolerance + spread;
    if (cost < best.cost - 1e-9) best = { slope, cost };
  }
  return Math.abs(best.slope) < 0.0026 ? 0 : best.slope;
}

const NUM_TAIL = /[\d,.]$/;
const NUM_HEAD = /^[\d,.]/;

function isCJK(ch) {
  return /[　-ヿ一-鿿＀-￯]/.test(ch);
}

/** 同一行内で近接するトークンを連結する（OCRの単語分割を戻す）。 */
function mergeRowTokens(row, medianHeight) {
  const sorted = [...row].sort((a, b) => a.x - b.x);
  const out = [];
  for (const raw of sorted) {
    const t = clone(raw);
    const prev = out[out.length - 1];
    if (!prev) {
      out.push(t);
      continue;
    }
    const gap = t.x - tokenRight(prev);
    const numericJoin = NUM_TAIL.test(prev.text) && NUM_HEAD.test(t.text);
    const cjkJoin = isCJK(prev.text[prev.text.length - 1]) && isCJK(t.text[0]);
    // OCRは1語を細かく割るため、日本語同士は文字高の0.8倍まで詰めて連結する
    const limit = numericJoin ? medianHeight * 0.9 : medianHeight * 0.8;
    if (gap >= -medianHeight * 0.3 && gap <= limit && (numericJoin || cjkJoin)) {
      prev.text += t.text;
      prev.w = tokenRight(t) - prev.x;
      prev.h = Math.max(prev.h || 0, t.h || 0);
      prev.conf = Math.min(prev.conf ?? 1, t.conf ?? 1);
      prev.merged = true;
    } else {
      out.push(t);
    }
  }
  return out;
}

const LABEL_AMOUNT = /^([^\d０-９]*[^\d０-９\s])[\s]*([-+−▲△¥￥(（]?[\d０-９][\d０-９,.\s]*[)）]?円?)$/;

/** 「基本給252,000」のようにラベルと金額がくっついたトークンを分割する。 */
function splitLabelAmount(token) {
  const text = token.text.trim();
  if (text.length < 3) return [token];
  const m = text.match(LABEL_AMOUNT);
  if (!m) return [token];
  const [, labelPart, amountPart] = m;
  if (!labelPart || labelPart.length < 2) return [token];
  if (!parseAmount(amountPart)) return [token];
  // 「2026年3月分」のような日付混じりは分割しない
  if (/[年月日]$/.test(labelPart)) return [token];
  const total = text.length;
  const w = token.w || 0;
  const labelW = (w * labelPart.length) / total;
  return [
    { ...token, text: labelPart, w: labelW, split: true },
    { ...token, text: amountPart, x: token.x + labelW, w: w - labelW, split: true },
  ];
}

/**
 * トークンを解析しやすい形へ整える。
 * @returns {{rows: Array<{index:number, tokens:Array, key:number}>, tokens:Array, slope:number, medianHeight:number}}
 */
export function buildLayout(rawTokens) {
  const tokens = (rawTokens || [])
    .filter((t) => t && typeof t.text === 'string' && t.text.trim() !== '')
    .map((t) => ({
      text: toHalfWidth(t.text).trim(),
      x: Number(t.x) || 0,
      y: Number(t.y) || 0,
      w: Number(t.w) || 0,
      h: Number(t.h) || 0,
      conf: t.conf === undefined ? 1 : Number(t.conf),
      page: t.page || 1,
    }))
    .filter((t) => t.text !== '');

  if (!tokens.length) {
    return { rows: [], tokens: [], slope: 0, medianHeight: 0 };
  }

  const medianHeight = medianOf(tokens.map((t) => t.h).filter((h) => h > 0)) || 10;
  const slope = estimateSkew(tokens, medianHeight);
  const tolerance = Math.max(2, medianHeight * 0.55);

  const grouped = clusterRows(tokens, slope, tolerance);
  const rows = [];
  for (const group of grouped) {
    const merged = mergeRowTokens(group, medianHeight).flatMap(splitLabelAmount);
    merged.sort((a, b) => a.x - b.x);
    rows.push({ tokens: merged, key: medianOf(merged.map((t) => rowKey(t, slope))) });
  }
  rows.sort((a, b) => a.key - b.key);
  rows.forEach((r, i) => {
    r.index = i;
    r.tokens.forEach((t) => {
      t.row = i;
    });
  });

  return {
    rows,
    tokens: rows.flatMap((r) => r.tokens),
    slope,
    medianHeight,
  };
}

/** 金額トークンの右端x座標から「金額列」を推定する。 */
export function detectAmountColumns(amountTokens, medianHeight) {
  const edges = amountTokens.map((a) => tokenRight(a.token)).sort((x, y) => x - y);
  const columns = [];
  const tol = Math.max(6, medianHeight * 1.2);
  for (const e of edges) {
    const last = columns[columns.length - 1];
    if (last && e - last.center <= tol) {
      last.members.push(e);
      last.center = last.members.reduce((s, v) => s + v, 0) / last.members.length;
    } else {
      columns.push({ center: e, members: [e] });
    }
  }
  return columns.filter((c) => c.members.length >= 2);
}

export { tokenRight, tokenCenterX, tokenCenterY };

File: native-ios/TedoriLogInput/Sources/TedoriLogCore/Layout.swift
import Foundation

/// 位置つきテキスト（PDF直接抽出でもVisionでも同じ形にする）。
/// 座標は左上原点・ポイント単位。
public struct TextToken: Codable, Equatable {
    public var text: String
    public var x: Double
    public var y: Double
    public var w: Double
    public var h: Double
    public var conf: Double
    public var page: Int

    public init(text: String, x: Double, y: Double, w: Double, h: Double, conf: Double = 1, page: Int = 1) {
        self.text = text
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.conf = conf
        self.page = page
    }

    public var right: Double { x + w }
    public var centerX: Double { x + w / 2 }
    public var centerY: Double { y + h / 2 }

    enum CodingKeys: String, CodingKey { case text, x, y, w, h, conf, page }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        text = try c.decode(String.self, forKey: .text)
        x = try c.decode(Double.self, forKey: .x)
        y = try c.decode(Double.self, forKey: .y)
        w = try c.decodeIfPresent(Double.self, forKey: .w) ?? 0
        h = try c.decodeIfPresent(Double.self, forKey: .h) ?? 0
        conf = try c.decodeIfPresent(Double.self, forKey: .conf) ?? 1
        page = try c.decodeIfPresent(Int.self, forKey: .page) ?? 1
    }
}

struct LayoutRow {
    var index: Int
    var tokens: [TextToken]
    var key: Double
}

struct Layout {
    var rows: [LayoutRow]
    var tokens: [TextToken]
    var slope: Double
    var medianHeight: Double
}

enum LayoutBuilder {

    static func median(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0 }
        let sorted = values.sorted()
        return sorted[sorted.count / 2]
    }

    private static func rowKey(_ token: TextToken, slope: Double) -> Double {
        token.centerY - slope * token.centerX
    }

    private static func clusterRows(_ tokens: [TextToken], slope: Double, tolerance: Double) -> [[TextToken]] {
        let sorted = tokens.sorted { rowKey($0, slope: slope) < rowKey($1, slope: slope) }
        var rows: [[TextToken]] = []
        var current: [TextToken] = []
        var anchor: Double?
        for token in sorted {
            let key = rowKey(token, slope: slope)
            if anchor == nil || abs(key - anchor!) <= tolerance {
                if anchor == nil { anchor = key }
                current.append(token)
                anchor = (anchor! * Double(current.count - 1) + key) / Double(current.count)
            } else {
                rows.append(current)
                current = [token]
                anchor = key
            }
        }
        if !current.isEmpty { rows.append(current) }
        return rows
    }

    /// 紙を斜めに撮った写真の傾きを推定する。
    static func estimateSkew(_ tokens: [TextToken], medianHeight: Double) -> Double {
        guard tokens.count >= 6 else { return 0 }
        let tolerance = max(2, medianHeight * 0.5)
        var best = (slope: 0.0, cost: Double.infinity)
        var slope = -0.08
        while slope <= 0.0801 {
            let rows = clusterRows(tokens, slope: slope, tolerance: tolerance)
            var spread = 0.0
            for row in rows {
                let keys = row.map { rowKey($0, slope: slope) }
                spread += (keys.max() ?? 0) - (keys.min() ?? 0)
            }
            let cost = Double(rows.count) * tolerance + spread
            if cost < best.cost - 1e-9 { best = (slope, cost) }
            slope += 0.0025
        }
        return abs(best.slope) < 0.0026 ? 0 : best.slope
    }

    private static func isCJK(_ ch: Character) -> Bool {
        guard let scalar = ch.unicodeScalars.first else { return false }
        switch scalar.value {
        case 0x3000...0x30FF, 0x4E00...0x9FFF, 0xFF00...0xFFEF: return true
        default: return false
        }
    }

    private static func isNumericTail(_ text: String) -> Bool {
        guard let last = text.last else { return false }
        return last.isNumber || last == "," || last == "."
    }

    private static func isNumericHead(_ text: String) -> Bool {
        guard let first = text.first else { return false }
        return first.isNumber || first == "," || first == "."
    }

    /// 同一行内で近接するトークンを連結する（OCRの単語分割を戻す）。
    private static func mergeRowTokens(_ row: [TextToken], medianHeight: Double) -> [TextToken] {
        let sorted = row.sorted { $0.x < $1.x }
        var out: [TextToken] = []
        for token in sorted {
            guard var prev = out.last else { out.append(token); continue }
            let gap = token.x - prev.right
            let numericJoin = isNumericTail(prev.text) && isNumericHead(token.text)
            let cjkJoin = isCJK(prev.text.last ?? " ") && isCJK(token.text.first ?? " ")
            let limit = numericJoin ? medianHeight * 0.9 : medianHeight * 0.8
            if gap >= -medianHeight * 0.3, gap <= limit, numericJoin || cjkJoin {
                prev.text += token.text
                prev.w = token.right - prev.x
                prev.h = max(prev.h, token.h)
                prev.conf = min(prev.conf, token.conf)
                out[out.count - 1] = prev
            } else {
                out.append(token)
            }
        }
        return out
    }

    private static let labelAmount = try! NSRegularExpression(
        pattern: #"^([^\d０-９]*[^\d０-９\s])[\s]*([-+−▲△¥￥(（]?[\d０-９][\d０-９,.\s]*[)）]?円?)$"#)

    /// 「基本給252,000」のようにラベルと金額がくっついたトークンを分割する。
    private static func splitLabelAmount(_ token: TextToken) -> [TextToken] {
        let text = token.text.trimmingCharacters(in: .whitespaces)
        guard text.count >= 3 else { return [token] }
        let range = NSRange(text.startIndex..., in: text)
        guard let m = labelAmount.firstMatch(in: text, range: range),
              let labelRange = Range(m.range(at: 1), in: text),
              let amountRange = Range(m.range(at: 2), in: text) else { return [token] }
        let labelPart = String(text[labelRange])
        let amountPart = String(text[amountRange])
        guard labelPart.count >= 2, Normalize.parseAmount(amountPart) != nil else { return [token] }
        if let last = labelPart.last, "年月日".contains(last) { return [token] }

        let labelWidth = token.w * Double(labelPart.count) / Double(text.count)
        var labelToken = token
        labelToken.text = labelPart
        labelToken.w = labelWidth
        var amountToken = token
        amountToken.text = amountPart
        amountToken.x = token.x + labelWidth
        amountToken.w = token.w - labelWidth
        return [labelToken, amountToken]
    }

    static func build(_ rawTokens: [TextToken]) -> Layout {
        let tokens = rawTokens.compactMap { token -> TextToken? in
            let text = Normalize.toHalfWidth(token.text).trimmingCharacters(in: .whitespaces)
            guard !text.isEmpty else { return nil }
            var copy = token
            copy.text = text
            return copy
        }
        guard !tokens.isEmpty else { return Layout(rows: [], tokens: [], slope: 0, medianHeight: 0) }

        let heights = tokens.map(\.h).filter { $0 > 0 }
        let medianHeight = median(heights) == 0 ? 10 : median(heights)
        let slope = estimateSkew(tokens, medianHeight: medianHeight)
        let tolerance = max(2, medianHeight * 0.55)

        var rows: [LayoutRow] = []
        for group in clusterRows(tokens, slope: slope, tolerance: tolerance) {
            let merged = mergeRowTokens(group, medianHeight: medianHeight).flatMap(splitLabelAmount)
            let sorted = merged.sorted { $0.x < $1.x }
            let key = median(sorted.map { rowKey($0, slope: slope) })
            rows.append(LayoutRow(index: 0, tokens: sorted, key: key))
        }
        rows.sort { $0.key < $1.key }
        for i in rows.indices { rows[i].index = i }

        return Layout(rows: rows, tokens: rows.flatMap(\.tokens), slope: slope, medianHeight: medianHeight)
    }

    /// 金額トークンの右端から「金額列」を推定する。
    static func detectAmountColumns(_ amounts: [Double], medianHeight: Double) -> [Double] {
        let edges = amounts.sorted()
        var columns: [(center: Double, members: [Double])] = []
        let tolerance = max(6, medianHeight * 1.2)
        for edge in edges {
            if var last = columns.last, edge - last.center <= tolerance {
                last.members.append(edge)
                last.center = last.members.reduce(0, +) / Double(last.members.count)
                columns[columns.count - 1] = last
            } else {
                columns.append((edge, [edge]))
            }
        }
        return columns.filter { $0.members.count >= 2 }.map(\.center)
    }

    /// ラベル列と金額列を、左から右の順序を保ったまま対応付ける。
    static func monotonicMatch(labelCenters: [Double], amountCenters: [Double]) -> [Int: Int] {
        let n = labelCenters.count
        let m = amountCenters.count
        guard n > 0, m > 0 else { return [:] }
        var dp = Array(repeating: Array(repeating: Double.infinity, count: m + 1), count: n + 1)
        var choice = Array(repeating: Array(repeating: "", count: m + 1), count: n + 1)
        for j in 0...m { dp[n][j] = 0 }
        for i in stride(from: n - 1, through: 0, by: -1) {
            for j in stride(from: m - 1, through: 0, by: -1) {
                let pair = abs(labelCenters[i] - amountCenters[j]) + dp[i + 1][j + 1]
                let skip = dp[i][j + 1] + 1
                if pair <= skip {
                    dp[i][j] = pair
                    choice[i][j] = "pair"
                } else {
                    dp[i][j] = skip
                    choice[i][j] = "skip"
                }
            }
        }
        var result: [Int: Int] = [:]
        var i = 0, j = 0
        while i < n && j < m {
            if choice[i][j] == "pair" {
                result[i] = j
                i += 1
                j += 1
            } else {
                j += 1
            }
        }
        return result
    }
}
